### ISSUE TITLE
Possible sub optimal blurriness

### ISSUE THREAD

**Issue by duyhung85** [kangax/fabric.js#4113]
Hi,

I am developing an editor based on fabricjs(2.0 beta3), I notice the blurriness on every objects in canvas. I go on to the main fabric website and experienced the same blurriness. Here is the screenshot: 

![screen shot 2017-07-20 at 12 32 37 pm](https://user-images.githubusercontent.com/4004056/28402175-aa66a104-6d48-11e7-8091-e7c02819af2c.png)

My screen is not retina and I check the fabric.devicePixelRatio value shows 1. The blurriness is very noticeable on my screen, but when I go to shutterstock editor (which is based on fabricjs) I dont see 
the blurriness when zoom browser in. So is this some fabricjs configuration I'm missing or any workaround on this matter?

Thanks,

**Comment by asturur** [comment]
i work on retina so i may miss some kind of blurriness. the website is still based on 1.7.16 but i m updating it today so that all problem will show up better. can you share a fiddle with your setup?

**Comment by duyhung85** [comment]
Sure, here is the fiddle (fabric 2.0 beta3):

[https://jsfiddle.net/duyhung85/58bfq8Ls/](url)

When I zoom in:
![screen shot 2017-07-20 at 3 22 04 pm](https://user-images.githubusercontent.com/4004056/28407769-7e5fbd2c-6d5f-11e7-9be3-e2393d42fcb8.png)

**Comment by samael205** [comment]
How do you do Zoom?

**Comment by asturur** [comment]
current master with no retina
![image](https://user-images.githubusercontent.com/1194048/28408373-8a3465da-6d37-11e7-8bcc-516011b66eeb.png)

1.7.15 no retina:
![image](https://user-images.githubusercontent.com/1194048/28408317-5ccbeb18-6d37-11e7-993a-4d01144b7bf9.png)

with the old 1.5 before cache
![image](https://user-images.githubusercontent.com/1194048/28408342-71d1714a-6d37-11e7-9289-952e741eb7ee.png)

apart the standard aliasing i do not see much difference.

Are you zooming a screenshot or are you zooming the page with css/ scrollwheel, zoom controls?

**Comment by asturur** [comment]
i may get sub optimal result being on a retina screen and disabling retina? screenshots are twice bigger for me and maybe some additional pixel manipulation occurs.

There are no fuzzyness issue known so far apart the standard subpixel cache placement

**Comment by asturur** [comment]
i agree that plain canvas makes a better job at painting text

![image](https://user-images.githubusercontent.com/1194048/28409027-a92c1508-6d39-11e7-9353-a550cb022fa0.png)

**Comment by duyhung85** [comment]
I dont know if this helping investigate the bug or not. I have to set 
`fabric.devicePixelRatio = 2;` 
after init the canvas in my application and it looks better.
Original:
![screen shot 2017-07-20 at 4 00 44 pm](https://user-images.githubusercontent.com/4004056/28409632-5d0cc5e2-6d65-11e7-9c91-dd79d3127d1d.png)

After manually set the fabric.devicePixelRatio = 2
![screen shot 2017-07-20 at 4 01 32 pm](https://user-images.githubusercontent.com/4004056/28409589-4195cdf4-6d65-11e7-9a78-7475a1b6c4d2.png)

Thanks

**Comment by asturur** [comment]
is strange, you are just oversampling, it should not happen.

**Comment by jasie** [comment]
I have a similar issue.

very sharp output of the "same" elements in KineticJS canvas but blurry output in Fabric.js canvas.

![fabricjs](https://user-images.githubusercontent.com/2750758/28459989-46a62a36-6e10-11e7-8b17-f6c6477cbcce.png)
![kineticjs](https://user-images.githubusercontent.com/2750758/28459990-46a79bb4-6e10-11e7-877e-be702c118462.png)

the buttons are rects with patternfill. the gray background behind the buttons is also a rect, will just grey fill and rx and ry.

fabric.devicePixelRatio does not change the output for me. the text is not blurry, though. strange. the big brown rectangle is a polygon.

fabric.devicePixelRatio = 2 does not change anything in my output.

**Comment by asturur** [comment]
Which is kinetic and which is fabricJS? can you attach both?

**Comment by asturur** [comment]
![image](https://user-images.githubusercontent.com/1194048/28460544-dacf8598-6e12-11e7-952a-fa102b0acaff.png)

this is the same comparision, cache, no cache, plain canvas but on a non retina screen.

I guess disabling retina no a retina screen makes worst results than a non retina screen

**Comment by jasie** [comment]
apologies, the fist one is the Fabric.js screenshot, the second one the KineticJS screenshot.

I also have this fabric.js-blurriness on desktop monitors.

**Comment by asturur** [comment]
Need some info, what are those icons? png? svg? single path?
How do you position them on the group?
On what fabricjs version you are and if you can make a blurry fiddle with a simplified example ( one icon on a rect is enough i guess )

**Comment by jasie** [comment]
Sure, the icons are png, in exactly the size of the rect (50x49px, funny designer...)
All rect elements are positioned in the group with group.getLeft() and group.getTop() and added with group.AddWithUpdate()
I'm on the latest fabric version (1.7.16).
I'll try to make a simple blurry fiddle - let's see, if I can reproduce the blurriness...

**Comment by asturur** [comment]
https://jsfiddle.net/yhp1m1tn/

@kennymoon does one of those text in the fiddle look sharper to you?

**Comment by Kennymoon** [comment]
![png](https://user-images.githubusercontent.com/32856767/31652077-e5f9259c-b358-11e7-846a-45b1b675a629.PNG)

2nd text looks shaper than others, buddy.

**Comment by asturur** [comment]
https://jsfiddle.net/yhp1m1tn/2/

Can you check if there are more sharp now?
Are they as sharp as 1.5?

**Comment by asturur** [comment]
With retina:
![image](https://user-images.githubusercontent.com/1194048/31652575-e15fc8f8-b31f-11e7-93ca-a80270eb1d05.png)

without retina for me
![image](https://user-images.githubusercontent.com/1194048/31652606-f8ed7b32-b31f-11e7-8ecb-3d06ec8813c9.png)

**Comment by ncou** [comment]
Same thing with my tests.
The reference 1.5 sample is more crisp.

**Comment by asturur** [comment]
Can you share a non retina version of it?

**Comment by HATyCEHKA** [comment]
The same fo me. String "Reference 1.5" is sharp, others are blur.
![image](https://user-images.githubusercontent.com/26141774/31706958-72d06012-b3f3-11e7-920b-42b324b76325.png)

**Comment by asturur** [comment]
So just to cheers you up, while i cannot fix it now on the live canvas, i have a different render system that makes it sharp for data url export. That means that if you have to create an asset from it, it will be sharp.

There is some positioning problem that is plaguing the text and everything else ( is just less noticeable ) and i have to find it.

But i have other things to do first.

**Comment by iongion** [comment]
Might be related, I have implemented zoom and I notice a border appears for all rectangles, even if they are set to have no border, this adds bluriness when in-fact the rects should have a sharp border.

![image](https://user-images.githubusercontent.com/867609/31774623-867d6210-b4e6-11e7-9444-d1cbf3696bff.png)

```
{
      left: 0,
      top: 0,
      width: 100,
      height: 100,
      evented: false,
      selectable: false,
      excludeFromExport: true,
      hasBorders: false,
      strokeWidth: 0,
      borderScaleFactor: 0,
}
```

Using `1.7.19`

**Comment by asturur** [comment]
yes that is the problem, either drawing bad on the cache or when positioning the cache on the canvas.
go figure out....

**Comment by asturur** [comment]
Just to be clear the cache has some inherited blurriness. once the text is drawn on the cache is aliased for that position, if you move it of half pixel, the aliasing will not be good for it. that we cannot solve.
But for when the position is exactly an integer number of pixel moved from the first draw, we should get crisp results.

**Comment by asturur** [comment]
Was just investigating a bit:

this is on screen
![image](https://user-images.githubusercontent.com/1194048/32694852-a2712f52-c74b-11e7-9852-9225c8ce82d4.png)

this is the cache
![image](https://user-images.githubusercontent.com/1194048/32694853-a85422bc-c74b-11e7-8591-87f14a6cfd17.png)

cache disabled, retina screen
![image](https://user-images.githubusercontent.com/1194048/32694859-cca7e02c-c74b-11e7-91b4-d95a40062aaa.png)

**Comment by asturur** [comment]
Ok i took another approach, i forgot for now the text, i started with line that are easy verifiable.

```javascript
var line = new fabric.Line([0, 3, 200, 3], {
stroke: 'black',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)
var line = new fabric.Line([0, 6, 200, 6], {
stroke: 'red',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)

var line = new fabric.Line([0, 9, 200, 9], {
stroke: 'black',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)
var line = new fabric.Line([0, 12, 200, 12], {
stroke: 'red',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)


// vertical



var line = new fabric.Line([3, 0, 3, 200], {
stroke: 'black',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)
var line = new fabric.Line([6, 0, 6, 200], {
stroke: 'red',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)

var line = new fabric.Line([9, 0, 9, 200], {
stroke: 'black',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)
var line = new fabric.Line([12, 0, 12, 200], {
stroke: 'red',
strokeWidth: 2,
noScaleCache: false,
})
canvas.add(line)
```
![image](https://user-images.githubusercontent.com/1194048/32698709-1d41209e-c7aa-11e7-8f05-089b6f8997ca.png)

I made so that the cache has a logic to try to draw always across roundy pixel, indipendently of the size of the obect.
So if the line is 1 pix, or the rect is 3 pixel wide, it will be drawn across half pixel so that it occupies 3 full pixels instead of 0.5 , 1 , 1 and 0.5 for example.
The result on the lines is good.

This is one pixel new code:
![image](https://user-images.githubusercontent.com/1194048/32698816-c16274f6-c7ab-11e7-846f-027f6441679e.png)


this is one pixel old code:
![image](https://user-images.githubusercontent.com/1194048/32698821-cb4b410a-c7ab-11e7-9fa9-3e6f05b1b415.png)

**Comment by asturur** [comment]
Text taken on a non retina screen cache disabled:
![image](https://user-images.githubusercontent.com/1194048/32698866-418eb716-c7ac-11e7-99fe-6481496fe5d0.png)

New code cache enabled ( apart for reference 1.5)
![image](https://user-images.githubusercontent.com/1194048/32698878-6849fd0c-c7ac-11e7-8ff0-77ad263aeb44.png)


Let's zoom:
on the left new code, center uncached, right cached.
The first 2 results are noticeably improved, the other are not and for now i cannot improve them.
The other results are fuzzy because the cache gets placed across pixels and there is no way to avoid fuzzyness without sacrifing precisions.

That would mean that a line with left 0, 0.1, 0.2, 0.3, 0.4, 0.49 would appear always in the same place and this is not possible.

![image](https://user-images.githubusercontent.com/1194048/32698917-41721010-c7ad-11e7-89db-fcdac10a8f02.png)

**Comment by asturur** [comment]
I merged what i achieved so far, please for who is still interested, fabricjs.com is updated with lastest code. give it a shot.

**Comment by HATyCEHKA** [comment]
2.0.0-rc.3 looks good!
All elements (text, lines, shapes and imported svg) are sharp.
Thanks a lot!

**Comment by nmorel** [comment]
@asturur Hello! You said `The other results are fuzzy because the cache gets placed across pixels and there is no way to avoid fuzzyness without sacrifing precisions.`
What can I do if I want to sacrifice precisions for sharper text ?

**Comment by asturur** [comment]
either disable cache, or you should position your text always in pixels.

zoom canvas with only integers, scale with only integers, use top and left with only integers. this should help.

**Comment by shaangidwani** [comment]
still same issue blurry text is render on canvas. try to set pixcel, object caching false but not getting result. waiting for the fix. 

http://jsfiddle.net/shaangidwani/91y5jb83/6/

**Comment by teuneboon** [comment]
This issue comes up pretty high on Google when searching for blurriness in fabricjs. So I'll add one more possible solution which solved it for me here.

If you have a css class on your canvas that sets a border or width/height you'll very likely get subtle blurriness. During development I added a 1px border on the canvas to see exactly where the canvas ended and had blurriness, when I removed the border it became sharp and many hacks I used to get it sharper weren't needed anymore.

Setting `imageSmoothingEnabled: false` on the canvas can also solve some issues with other types of blurriness.
